### ISSUE TITLE
`--emit metadata` emitting empty .rmeta file

### ISSUE THREAD

**Issue by 0dvictor** [rust-lang/rust#67293]
At present, `rustc --emit=metadata foo.rs` generates an empty rmeta file named `libfoo.rmeta`. The logic behind is that [rmeta is not needed for certain crate types](https://github.com/rust-lang/rust/blob/12307b3b08edee543a78fb9d4a837fbd6d6ac0fa/src/librustc_interface/passes.rs#L922-L939). However, I feel we should still save it to a file when user explicitly requested it by `--emit=metadata`, especially the metadata is always available in memory regardless of output crate types.

**Comment by ehuss** [comment]
I would be cautious with this, as it could be a noticeable increase in disk usage for some projects.  Rust already uses too much disk space.  There should probably be a stronger reason than "it seems like it should".

**Comment by bjorn3** [comment]
Emitting an empty .rmeta file makes sense, as the metadata is what is needed to use it with `--extern`/`extern crate`. For executables, cdylibs and staticlibs that is impossible, so no metadata is needed in those cases.

**Comment by 0dvictor** [comment]
> There should probably be a stronger reason than "it seems like it should".

Definitely. IMO, user's explicit request should be honored; therefore, in this case, the explicit request is `---emit=metadata`.

On the other side, I agree with @bjorn3 's point that the metadata is not needed for certain crate types; however, under this scenario, I feel rustc should print an error or warning message instead of silently generating an empty file.

**Comment by 0dvictor** [comment]
> it could be a noticeable increase in disk usage for some projects

Does it mean these projects request `--emit=metadata` with crate types that do not need metadata? If it does, is there any obstacles or restrictions that these project cannot remove `--emit=metadata` from the arguments?

**Comment by bjorn3** [comment]
Does `cargo check` use `--emit=metadata` for executables?

**Comment by ehuss** [comment]
> Does it mean these projects request `--emit=metadata` with crate types that do not need metadata?

Yes. this is how `cargo check` works.

**Comment by cosmicexplorer** [comment]
@ehuss 

> I would be cautious with this, as it could be a noticeable increase in disk usage for some projects. Rust already uses too much disk space. There should probably be a stronger reason than "it seems like it should".

*cargo* uses too much disk space, because *cargo* does not have any transitive checksum-based invalidation model employing something like LRU gc, so *cargo* is for example unable to avoid repeatedly regenerating said `.rmeta` file. the proposal in OP is the one single thing i would need from rustc in order to implement the kind of extremely powerful cached *and* pipelined build process we implemented for scala compiles with pants https://github.com/bazelbuild/rules_rust/issues/428#issuecomment-3695407768

**Comment by cosmicexplorer** [comment]
Reading about the query-based architecture of the rust compiler just now and it being motivated by incrementality and am very surprised to realize how strong the analogy is to the pants v2 engine (https://rustc-dev-guide.rust-lang.org/queries/incremental-compilation.html#the-try-mark-green-algorithm).

@jyn514 for example the bootstrapping "mini pants" task system I was pondering for a prototype to unify `bootstrap` and `cargo` would have been making extremely similar tradeoffs as described for the rustc incremental compilation system https://rustc-dev-guide.rust-lang.org/queries/incremental-compilation-in-detail.html#query-modifiers and since I want it to be cute (unlike pants) I was absolutely going to have an explicitly-invoked content-addressed store in the filesystem.

(& btw If anyone knows offhand where I should look to learn more about this incrementality persistence mechanism plz lmk https://rustc-dev-guide.rust-lang.org/queries/incremental-compilation-in-detail.html#incrementality-of-on-disk-data-structures)

# zip-based multiple crate archive format proposal

But the really important thing I wanted to mention was that reading enough of @bjorn3's posts today caused me to feel like I had an epiphany of sorts but maybe not:

(a) we (often) actually lose information when we generate the system-native archive format, particularly the ability to split back out into unambiguous record boundaries
@bjorn3 https://github.com/rust-lang/rust/issues/73632#issuecomment-1116696323
> By the way I forgot if I asked this already, but would a rustc command to bundle multiple rlibs into a single staticlib without requiring any rust source for this staticlib work? So something like `rustc --crate-type staticlib -Zlink-libs liba.rlib libb.rlib -o libcombined.a`. This should allow diamond dependencies I think. You could run it just before invoking the system linker.

(b) in addition to crate-specific data/config, there is global data/config we need before we can generate an output binary. it seems that we lose the ability to modify that information once we bounce to the system-native format
@bjorn3 https://github.com/rust-lang/rust/issues/73632#issuecomment-1719158688
> We already have the staticlib crate type for bundling everything into a single `.a` file. Turning individual rlibs into `.a` files won't work as it would either duplicate the allocator shim and such between every such `.a` or or omit them from all and produce unlinkable `.a` files. One idea I have is to produce a new crate type which is to `staticlib` as `dylib` is to `cdylib`. It would act like a `dylib` with respect to producing the allocator shim and bundling multiple crates, but produce a `.a` file instead of a `.so` file and add the necessary crate metadata to allow consuming it like a regular crate from rustc. See also the end of [#111594 (comment)](https://github.com/rust-lang/rust/issues/111594#issuecomment-1550087206)

(c) the "crate type" seems ridiculously overloaded and it seems confusing by itself to me that it can represent a single (?) compiled crate or like multiple
@bjorn3 https://github.com/rust-lang/rust/issues/111594#issuecomment-1550087206
> A couple of days ago my PR [#106560](https://github.com/rust-lang/rust/pull/106560) got merged. By combining `-Z staticlib-allow-rdylib-deps` with `-Z staticlib-prefer-dynamic`, libstd will be dynamically linked and `--print native-static-libs` will tell you how to dynamically link against libstd. This does not however enable statically linking libstd. I'm not sure what the best way to add that is though.
> 
> The `staticlib` just like `cdylib` is not supposed to export any symbols outside of those that are `#[no_mangle]`, which would make `libstd.a` impossible. Currently it does export all symbols as there is no way to indicate that a symbol shouldn't be exported from an archive (for `cdylib` we have control over the linker, for `staticlib` we don't). This needs to be fixed to allow multiple rust `staticlib`'s to be linked into the same executable or dynamic library without getting symbol conflicts. ([#104707](https://github.com/rust-lang/rust/issues/104707)) Maybe one option would be to have something like a `rstaticlib` crate type which can bundle multiple crates and can be statically linked (that is it is to `staticlib` what `dylib` is to `cdylib`) and then additionally ship libstd as `rstaticlib`.

& then this part:
> Maybe one option would be to have something like a `rstaticlib` crate type which can bundle multiple crates

## affordances of a zip-based CAS archive 
I am *not* an expert in backend codegen, but I have spent immense amounts of time with the zip file format (the `zip` crate will never have a CVE on my watch but we *do* have SIMD parsing because of me) & with packaging software. In python, we have _distributions_ (which I analogize to the current set of rust _crate types_), but we *also* have _zipapps_, which are themselves actually much less structured than [pex files](https://docs.pex-tool.org/).

Also important: [PEP 777](https://peps.python.org/pep-0777/) is the wheel deal (and Emma Smith who is authoring it is *also* the author of the Rust PEP and she rocks). See [my infodump in the PEP thread](https://discuss.python.org/t/pep-777-how-to-re-invent-the-wheel/67484/247).

Some final points:
1. both java and python allow loading code from a zip file. for various reasons, this can *sometimes but not always* be significantly faster than traversing the filesystem, particularly in the case of generated module paths with many many entries.
    - it also means you can actually be 100% confident about where your code is being loaded from (death to linear path traversal!!!!)
2. so rust already has the [crate hash](https://rustc-dev-guide.rust-lang.org/backend/libs-and-metadata.html#strict-version-hash). python's analogy would be the [direct url](https://packaging.python.org/en/latest/specifications/direct-url-data-structure/). please consider the following:
    - https://github.com/pantsbuild/pants/pull/8793 => you can safely load a resource from from a registered external store (analogy to https://rustc-dev-guide.rust-lang.org/query.html#how-queries-interact-with-external-crate-metadata)
    - https://pip.pypa.io/en/latest/reference/installation-report/#specification => PEP 751 maliciously erased my years of effort to get consensus on this standard i created so every python user could rely upon a standard to do my `--generate-ipex` work
      - (which i developed specifically to serve twitter data scientists with slow connections to our US-based data centers)
    - so pex actually has a filesystem-based CAS store which it unconditionally extracts to upon first startup. but you can actually use a zip itself as a store without decompression https://github.com/pex-tool/pex/pull/2175
3. so i need to actually implement this, but zstd dicts can be used to get faster-than-light compression ratios by codifying context: https://codeberg.org/cosmicexplorer/corporeal
    - will be proposing this as part of a wheel format at some point, but unfortunately the PEP process and reference impl for python packaging are hostile territory atm https://github.com/pypa/packaging/issues/570#issuecomment-1690498086
    - i also plan to start work on a rust-native zstd impl this week
4. it's kind of a gimmick but you can always just append to the end of a zip file's local entries (before the central directory) and this makes it *incredibly* useful for incremental accumulation of structured persistent data
    - i realized you can both extract zips in parallel as well as construct them this way via recursively merging without intermediate decompression: https://docs.rs/zip/latest/zip/write/struct.ZipWriter.html#method.merge_archive
5. astral blatantly steals credit for this technique i invented too but basically zip archives can be introspected even over a network https://github.com/pypa/pip/issues/7819
    - this is the absolutely bulletproof version https://github.com/pypa/pip/pull/12208 but the sketch is
> _hacky_extract_sub_reqs() (see https://github.com/cosmicexplorer/pip/blob/a60a3977e929cfaed6d64b0c9e3713d7c502e51e/src/pip/_internal/resolution/legacy/resolver.py#L550-L552) will:
> a. send a HEAD request to get the length of the zip file
> b. perform several successive GET requests to extract the relative location of the METADATA file
> c. extract the DEFLATE-compressed METADATA file and INFLATE it
> d. parse all Requires-Dist lines in METADATA for requirement strings

@bjorn3:
> Furthermore there may be performance benefits to using different archive formats
this is rust!!!!! we didn't just stick with null terminated strings because they were "system native"!!!!!

at a basic level, differentiating "flexible archive format we can extend for our own purposes" seems SUPER important to differentiate from "system-native archive format we can provide to a C linker".

i unfortunately have absolutely no clue what `ar` or `ld` does to generate a `.a` archive (yet), but i am absolutely 10000% sure that generating a zip-based `rstaticlib` absolutely *can* be pipelined, and that a zip is much more structured than most other types of formats i can name

oh and i'm greedy too this means my rust build system can incorporate a concept of ABI and target compatibility too

@bjorn3  https://github.com/rust-lang/rust/issues/67293#issuecomment-565726479
> Emitting an empty `.rmeta` file makes sense, as the metadata is what is needed to use it with `--extern`/`extern` crate. For executables, cdylibs and staticlibs that is impossible, so no metadata is needed in those cases.

thanks for this explanation of those semantics! do i understand correctly, then, that your proposed new output type `rstaticlib`  would be a way to e.g. enforce the emission of a `.rmeta` while retaining backwards compat?

thx for listening!! i will probably put this up as a pre-RFC on the discourse tomorrow!
